src/uct/tcp/tcp_ep.c

```diff
@@ -20,25 +20,30 @@
                                    _am_arg, ...) \
     _target_length = _pack_f(_target_buf, _am_arg)
 
-#define UCT_TCP_AM_PREPARE(_ep, _id, _len_thr, _hdr, \
-                           _pack_f, _am_payload, _payload_length, \
-                           _am_header, _method, _name) \
+#define UCT_TCP_AM_PREPARE(_iface, _ep, _id, _hdr, _pack_f, _am_payload, \
+                           _payload_length, _am_header, _method) \
     do { \
         UCT_CHECK_AM_ID(_id); \
         \
         if (!uct_tcp_ep_can_send(_ep)) { \
+            UCS_STATS_UPDATE_COUNTER((_ep)->super.stats, UCT_EP_STAT_NO_RES, 1); \
+            return UCS_ERR_NO_RESOURCE; \
+        } \
+        \
+        ucs_assertv((_ep)->tx.buf == NULL, "ep=%p", _ep); \
+        \
+        (_ep)->tx.buf = ucs_mpool_get_inline(&(_iface)->buf_mpool.tx); \
+        if (ucs_unlikely((_ep)->tx.buf == NULL)) { \
+            UCS_STATS_UPDATE_COUNTER((_ep)->super.stats, UCT_EP_STAT_NO_RES, 1); \
```

Comment:
most of this macro can be extracted to inline function which checks all conditions and returns OK/error

Reply:
refactored